study1/reverse-words-in-a-string-iii.java

```diff
@@ -0,0 +1,37 @@
+class Solution {
+    public String reverseWords(String s) {
+        //check if length of s is 1 or s is null
+        //s.length() is an operation that returns the length of the string s,
+        //so it takes O(1) time to check the length
+        if (s.length() <= 1 || s == null) {
+            return s;
+        }
+        //split() -> performed proportionally to the length of the string
+        //if the length of String is n, so it takes O(n) time to split the String
+        //size of this splited array is proportional to the number of words in input String
+        //so the size of the splited array is O(n).
+        String[] splitedArr = s.split(" ");
+        //create new reversed Array of input String
+        //size of this array is proportional to the number of words in splited string array
+        //so the size of the reversed array is O(n).
+        String[] reversedArr = new String[splitedArr.length];
+
+        for (int i = 0; i < splitedArr.length; i++) {
+            //create a Stringbuffer to use the reverse function
+            //size of this sb is proportional to the length of each word 
+            //and it depends on the maximum word length
+            //if length of maximum word is m -> size of sb is O(m)
+            StringBuffer sb = new StringBuffer(splitedArr[i]);
+            //add the inverted String to the new String Array
+            //reversing each word takes time proportional to the length of the word
+            //if there is n words in the Strings -> it takes O(n) time to reverse the word
+            reversedArr[i] = sb.reverse().toString();
```

Comment:
This doesn't affect the time complexity of the bigO notation, but basically its doing operation twice than it was needed.
1. Copying into the string buffer `new StringBuffer(splittedArr[i])`
2. And then reversing it.

You can avoid this duplication by copying in while reversing it. then it'll take less operation.

Reply:
I've been thinking about it for a long time, but I don't know still how to avoid duplication. Isn't that just creating a stringbuffer and reversing it??😣 how.. how can.. avoid duplication.. 🧐

Reply:
Sorry, when I said duplication, I just meant the number of operations being performed, thus I actually meant unnecessary. 
So, instead of the following
```
for (int i = 0; i < splitedArr.length; i++) {
            StringBuffer sb = new StringBuffer(splitedArr[i]); // Copy in n characters --> n operations
            reversedArr[i] = sb.reverse().toString(); // reverse n characters --> n operations
}
```

You can do 
```
for (int i = 0; i < splitedArr.length; i++) {
            StringBuffer sb = new StringBuffer(splitedArr[i].length); // Reserve n characters --> 1 operation
            for(int j = splitedArr[i].length - 1; j >= 0 ; --j ) { // copy n characters --> n operation.
               sb.append(splitedArr[i][j]); 
            }
}
```
and instead of doing 2n operations. you only do n operation.